pkg/products/rhssouser/reconciler.go

```diff
@@ -243,9 +268,342 @@ func (r *Reconciler) reconcileComponents(ctx context.Context, installation *inte
+	}
+
+	// Create / update the synchronized users
+	for _, user := range users {
+		or, err = r.createOrUpdateKeycloakAdmin(user, installation, ctx, serverClient)
+		if err != nil {
+			return integreatlyv1alpha1.PhaseFailed, errors.Wrap(err, "failed to create/update the customer admin user")
+		} else {
+			r.logger.Infof("The operation result for keycloakuser %s was %s", user.UserName, or)
+		}
+	}
+
 	return integreatlyv1alpha1.PhaseCompleted, nil
 }
 
+func identityProviderExists(kcClient keycloakCommon.KeycloakInterface) (bool, error) {
+	provider, err := kcClient.GetIdentityProvider(idpAlias, "master")
+	if err != nil {
+		return false, err
+	}
+	if provider != nil {
+		return true, nil
+	}
+	return false, nil
+}
+
+// The master realm will be created as part of the Keycloak install. Here we update it to add the openshift idp
+func (r *Reconciler) updateMasterRealm(ctx context.Context, serverClient k8sclient.Client, installation *integreatlyv1alpha1.Installation, host string) (*keycloak.KeycloakRealm, error) {
+	keycloakMasterRealmName := "master"
+
+	kcr := &keycloak.KeycloakRealm{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      keycloakMasterRealmName,
+			Namespace: r.Config.GetNamespace(),
+		},
+	}
+
+	or, err := controllerutil.CreateOrUpdate(ctx, serverClient, kcr, func() error {
+		kcr.Spec.RealmOverrides = []*keycloak.RedirectorIdentityProviderOverride{
+			{
+				IdentityProvider: idpAlias,
+				ForFlow:          "browser",
+			},
+		}
+
+		kcr.Spec.InstanceSelector = &metav1.LabelSelector{
+			MatchLabels: getInstanceLabels(),
+		}
+
+		kcr.Labels = getMasterLabels()
+
+		kcr.Spec.Realm = &keycloak.KeycloakAPIRealm{
+			ID:          keycloakMasterRealmName,
+			Realm:       keycloakMasterRealmName,
+			Enabled:     true,
+			DisplayName: keycloakMasterRealmName,
+		}
+
+		// The identity providers need to be set up before the realm CR gets
+		// created because the Keycloak operator does not allow updates to
+		// the realms
+		err := r.setupOpenshiftIDP(ctx, installation, kcr, serverClient, "master", host)
+		if err != nil {
+			return errors.Wrap(err, "failed to setup Openshift IDP for user-sso")
+		}
+
+		return nil
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to create/update keycloak realm: %w", err)
+	}
+	r.logger.Infof("The operation result for keycloakrealm %s was %s", kcr.Name, or)
+
+	return kcr, nil
+}
+
+func (r *Reconciler) createOrUpdateKeycloakAdmin(user keycloak.KeycloakAPIUser, inst *integreatlyv1alpha1.Installation, ctx context.Context, serverClient k8sclient.Client) (controllerutil.OperationResult, error) {
+	selector := &keycloak.KeycloakUser{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      fmt.Sprintf("generated-%v", user.UserName),
+			Namespace: r.Config.GetNamespace(),
+		},
+	}
+
+	return controllerutil.CreateOrUpdate(ctx, serverClient, selector, func() error {
+		ownerutil.EnsureOwner(selector, inst)
```

Comment:
Installation resource should not own a KeycloakUser.
Installation CR is namespaced, and resides in a different namespace then KeycloakUser.
Such owner reference is invalid as per Kubernetes design. See [garbage collection docs](https://kubernetes.io/docs/concepts/workloads/controllers/garbage-collection/).
Relevant piece:
```
Note: Cross-namespace owner references are disallowed by design. This means: 1) Namespace-scoped dependents can only specify owners in the same namespace, and owners that are cluster-scoped. 2) Cluster-scoped dependents can only specify cluster-scoped owners, but not namespace-scoped owners.
```

I now see that such invalid owner reference exists in a couple of places in SSO reconcilers.
We can create a follow up ticket, you don't have to fix all of them now.

Reply:
Follow up issue created here - https://issues.redhat.com/browse/INTLY-5168